### PR TITLE
Enable replicated enforcement and implement broadcast execution

### PR DESCRIPTION
**Pull request by arzuschen**
This patch consists of two main parts:
1) Enable replicated enforcement
Now replicated distribution can be enforced from distribution on any list expression through broadcast node. This make it possible for hashjoining two distributed tables by broadcasting the smaller table for the build end. Therefore the total cost may decrease since a join previously requiring two redistribution can now be replaced by only one broadcast.
2) Implement broadcast execution
Broadcast physic node execution is implemented using shared code from redistribution. The class of Physic RemoteExchange is modified to adopt to the shared portion for broadcast and redistribution. The estimated cost for broadcast is set to the same as redistribution, which is 2.0x the cardinality of child node.

After the above is implemented, the plans for redistribution unittest changes since using broadcasting is likely to produce lower cost plans. Therefore, a parameter in option.optimize is introduced to turn on/off the replicated distribution enforcement. This way the plans for redistribution test does not need to change.
Distributed TPCH result plans are changed accordingly.

### CODE REVIEW

File: qpmodel/PhysicalNode.cs [hunk 1]
```diff
@@ -1904,8 +1907,47 @@ public PhysicNode EmulateSerialization(PhysicNode node)
             var newnode = node.Clone();
             return newnode;
         }
-        public virtual string OpenConsumer(ExecContext context) => null;
-        public virtual string OpenProducer(ExecContext context) => null;
+        public virtual string OpenConsumer(ExecContext econtext)
```
Comment (zhouqingqing) [inline qpmodel/PhysicalNode.cs:1910]: I would suggest you keep OpenConsumer() as it is but provides an helper function say createProducersAndChannels() for Redistribute/Broadcast to reuse.

Same as other virtual functions.

Comment (arzuschen) [reply qpmodel/PhysicalNode.cs:1910]: got it

File: test/UnitTest.cs [hunk 2]
```diff
@@ -2256,73 +2256,75 @@ public void Gather()
         [TestMethod]
         public void Redistribute()
         {
+            var option = new QueryOption();
+            option.optimize_.enable_broadcast_ = false;
```
Comment (zhouqingqing) [inline test/UnitTest.cs:2260]: how about adding a loop on turn on/off enable_broadcast_ and assert either way the results are the same. If the plan is different, assert the difference. See similar usage in tpch tests.

File: test/regress/expect/tpch0001_d/q16.txt [hunk 3]
```diff
@@ -28,33 +28,33 @@ order by
 	p_brand,
 	p_type,
 	p_size
-Total cost: 6300.38, memory=22274
-PhysicOrder  (inccost=6300.38, cost=754.38, rows=148, memory=6364) (actual rows=0)
+Total cost: 4774.38, memory=22274
```
Comment (zhouqingqing) [inline test/regress/expect/tpch0001_d/q16.txt:31]: good - at least we see with broadcast in, all new cost are down!

File: test/UnitTest.cs (hunk 2 above)
Comment (arzuschen) [reply test/UnitTest.cs:2260]: sure, can I create a broadcast test method to do that?

**Pull request merged**